aws-http-auth/sigv4a/credentials.go

```diff
@@ -0,0 +1,179 @@
+package sigv4a
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	"hash"
+	"math"
+	"math/big"
+	"sync"
+
+	"github.com/aws/smithy-go/aws-http-auth/credentials"
+)
+
+var (
+	p256          elliptic.Curve
+	nMinusTwoP256 *big.Int
+
+	one = new(big.Int).SetInt64(1)
+)
+
+func init() {
+	p256 = elliptic.P256()
+
+	nMinusTwoP256 = new(big.Int).SetBytes(p256.Params().N.Bytes())
+	nMinusTwoP256 = nMinusTwoP256.Sub(nMinusTwoP256, new(big.Int).SetInt64(2))
+}
+
+// ecdsaCache stores the result of deriving an ECDSA private key from a
+// shared-secret identity.
+type ecdsaCache struct {
+	mu sync.Mutex
+
+	akid string
+	priv *ecdsa.PrivateKey
+}
+
+// Derive computes and caches the ECDSA key-pair for the identity, returning
+// the result.
+//
+// Future calls to Derive with the same set of credentials (identified by AKID)
+// will short-circuit. Future calls with a different set of credentials
+// (identified by AKID) will re-derive the value, overwriting the old result.
+func (c *ecdsaCache) Derive(creds credentials.Credentials) (*ecdsa.PrivateKey, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if creds.AccessKeyID == c.akid {
+		return c.priv, nil
+	}
+
+	priv, err := derivePrivateKey(creds)
+	if err != nil {
+		return nil, err
+	}
+
+	c.akid = creds.AccessKeyID
+	c.priv = priv
+	return priv, nil
+}
+
+// derivePrivateKey derives a NIST P-256 PrivateKey from the given IAM
+// AccessKey and SecretKey pair.
+//
+// Based on FIPS.186-4 Appendix B.4.2
+func derivePrivateKey(creds credentials.Credentials) (*ecdsa.PrivateKey, error) {
```

Comment:
I'll start taking a look at this compared with the docs you referenced, but I'll take time and I'm concerned since these are crypto functions that I'm not familiar with. Looking at what other SDKs do, they basically import the CRT and call it a day, which we don't want to do

* [.NET](https://github.com/aws-samples/sigv4a-signing-examples/blob/main/dotnet/SigV4A/SigV4ASign.cs#L1)
* [Python](https://github.com/aws-samples/sigv4a-signing-examples/blob/main/python/sigv4a_sign.py#L2)
* [Java](https://github.com/aws-samples/sigv4a-signing-examples/blob/main/java/src/main/java/com/sigv4aSigning/SigV4ASign.java#L5)

We can test correctness by "we were able to call the service and it works" but we may need some extra eyes on this. Curious to hear your thoughts on it

Reply:
FWIW, this derivation component is ripped from the internal go v2 implementation, which previously had an AppSec review.

Reply:
But as far as testing correctness goes - yeah, not sure we can do any better than we were doing before in the SDK. In fact we're technically probably testing it _better_ now since we have a dedicated e2e test for it.